wg-api/spec-documents/electron-c-apis.md

```diff
@@ -0,0 +1,161 @@
+
+## Summary
+
+Similar to the [Node-API (also known as N-API) of Node.js](https://nodejs.org/api/n-api.html), the Electron C APIs is a set of C APIs providing access to Electron and Chromium’s resources, which can be used in Node.js native modules.
+
+## Platforms
+
+All
+
+## Impetus
+
+There are 2 reasons:
+
+### Easier to add new APIs
+
+The majority of Electron’s JavaScript APIs are about accessing native resources of the operating system; some of them could not be implemented as native modules because they need to interact with Electron and Chromium’s resources.
+
+For example, the `TouchBar` and `inAppPurchase` APIs must live as built-in APIs because the JavaScript bindings must interact with internal `NSApp` and `NSWindow`, and there is no non-intrusive way doing that outside Electron.
+
+With Electron C APIs we can make it possible to have those APIs as native modules. This can make Electron less bloating since we don't have to make every new API a built-in API , and for users who want to add a large set of new APIs they can now only request a few minimal C APIs in Electron and then implement anything they want in native modules.
+
+### Reduce Electron forks
+
+Some people may already know that the official releases of VS Code and Teams use custom builds instead of the offical one, this is because they have some modifications that can not be upstreamed (for example some hacks to the `<video>` tag). By exposing certain internal Chromium APIs as Electron C APIs, we can make those modifications live as native modules instead of having a custom build.
+
+## API Design
+
+__Note that this spec does not define any speicific API and the C APIs used in this spec are only for demonstration, the actual APIs should be discussed in follwoup PRs.__
+
+The Electron C APIs should be very cheap: they only take minimal disk space, usually less than 40 bytes per API since they are just thin wrappers; and they have no runtime cost.
+
+The Electron C APIs should need no API desgin: they are mostly direct C wrappers of C++ interface.
+
+The Electron C APIs should be easy to pass API review: if there is an actual usage we should be able to approve a C API quickly, since they are usually cheap and exposing an existing API.
+
+To use the Electron C APIs in native modules, users only need to include the related headers:
+
+```c
+#include <node_api.h>
+#include <electron/version.h>
+#include <electron/main/app.h>
+#include <electron/main/base_window.h>
+#include <electron/common/native_image.h>
+#include <chromium/main/view.h>
+#include <chromium/renderer/render_frame.h>
+
+// Get the native window handle of Electron window.
+NSWindow* window = electron_base_window_get_native_handle(object);
+[window setToolbar:toolbar];
+
+// Create native views.
+electron_view_t view = electron_view_create();
+electron_button_t button = electron_button_create("Title");
+electron_view_add_child_view(view, button);
+electron_window_set_content_view(window, view);
```

Comment:
What schema do you imagine for anything that needs to subscribe to events?

These kinds of objects currently have a V8-managed lifetime. How should the C APIs expose GC-related primitives? Does each C handle map to a V8 handle under the hood, so that the objects remain referenced and don't get garbage collected? Or do we provide some other system for C APIs to interact with GC (e.g. refcounting)?